### ISSUE TITLE
Consider integration with Paging library

### ISSUE THREAD

**Issue by Zhuinden** [realm/realm-java#5486]
See https://medium.com/@Zhuinden/live-paged-lists-architecture-components-and-room-or-realm-268c9299a661

**Comment by Zhuinden** [comment]
I've talked to @ericmaxwell2003 and he said he'll look into it later.

**Comment by aperfilyev** [comment]
Paging is 1.0 as of may 8th

**Comment by Zhuinden** [comment]
Yeah I have to do it here too https://github.com/Zhuinden/realm-monarchy/issues/3

**Comment by mhdtouban** [comment]
Any plans to support paging?!

**Comment by Zhuinden** [comment]
Slightly experimental, but [Monarchy](https://github.com/Zhuinden/realm-monarchy) supports it now.

``` java
 DataSource.Factory<Integer, RealmDog> realmDataSourceFactory = monarchy.createDataSourceFactory(realm -> realm.where(RealmDog.class)); 
 dataSourceFactory = realmDataSourceFactory.map(input -> Dog.create(input.getName())); 
 dogs = monarchy.findAllPagedWithChanges(realmDataSourceFactory, 
                                         new LivePagedListBuilder<>(dataSourceFactory, 20)); 
```

> (not having to explicitly open/close, and while keep the notification could be moved to a background handler thread

Well would you look at that, that's what Monarchy does 😄

**Comment by mhdtouban** [comment]
@cmelchior I want to use paging for many reasons, first we now officially have a way for pagination in Android instead of relying on recycling view scroll listener which has downsides, second paging offers cool features like a placeholder and continuous loading mechanism etc.

**Comment by cmelchior** [comment]
Note, It isn't that we don't want to support paging, but I strongly suspect that the majority of cases don't actually need it and people are just using patterns from elsewhere that doesn't really apply to Realm.

**Comment by mhdtouban** [comment]
Am not quite sure I get you with "don't need those loading mechanisms with realm", let's say I want to fetch data through an API with pagination. How would realm help in lazy loading? I mean I understand that it will help if am hitting the database, but what if am getting the data from API?

**Comment by Zhuinden** [comment]
Personally I started investigating the background thread read because of [this] (https://github.com/realm/realm-java/issues/5328). `realm.refresh()` is a bit less free than it used to be.

> "don't need those loading mechanisms with realm", let's say I want to fetch data through an API with pagination. How would realm help in lazy loading? I mean I understand that it will help if am hitting the database, but what if am getting the data from API?

Technically you get the benefits of BoundaryCallback being called for zero items loaded and for item at end loaded. Previously you had to intercept zero-loaded in the Results change listener, and you had to know you reached the end if RecyclerView could no longer scroll down. 

BoundaryCallback does wrap these cases.

**Comment by cmelchior** [comment]
@mhdtouban Sorry, I was mainly thinking about the use case where you had all data stored locally. I do agree that the paging library nicely provides a lot of utility where you need to fetch the data from the server first.

**Comment by ozzioma** [comment]
Guys great work!

Can I suggest examples explaining how exactly pagination is not needed with Realm?
Sorry but this is not sufficient

> Since queries in Realm are lazy, performing this sort of paginating behavior isn’t necessary at all, as Realm will only load objects from the results of the query once they are explicitly accessed.
> 
> If for UI-related or other implementation reasons you require a specific subset of objects from a query, it’s as simple as taking the IQueryable object, and reading out only the objects you need.
> 

I mean this problem of pagination with Realm gets most new users stumped, and there's no example!
Honestly it's a hard concept telling users to replace well known pagination routines with a loop...

Most of the UI controls and grid/list display routines out there have callbacks that include query, sort parameters...**and pagination parameters**.
A simple example would have clarified this feature of not needing "those loading mechanisms with realm".
Would it be too much work to add a few lines of examples to your documentation showing exactly how it's done? 

It's clear the use case being referenced in your documentation is not sufficient to cater to the majority of pagination use cases out there. 

**You do need a way to execute SKIP, TAKE, TOP, LIMIT routines.** 
If Realm already does that for local data, can we see how?

Thanks.

**Comment by Zhuinden** [comment]
Technically what they mean by not needing pagination is that data is lazy-loaded so that means you can access any item at any index, and that item will only be loaded on access. This lazy loading means that you don't need to "load more" because you can just throw everything into the list regardless of result size, and it'll be fine.

But Realm doesn't actually support "SKIP.

**Comment by BapNesS** [comment]
Hi everyone,

To be consistent and use support libraries, I'm wanting to implement Paging support library because I need to fetch data from the API when needed.
`LivePagedListBuilder` require `DataSource` and `PagedList.BoundaryCallback` implementations.

Because previous comments are 5 months old, I just wanted to know if an official `RealmDataSource` is on a way or not at all ?

I think it would be nice to have one, like ObjectBox do for example.

Thanks.

![Paging-01](https://user-images.githubusercontent.com/4005409/54219508-f2a0ae80-44ef-11e9-9605-44a4daf401d2.png)

**Comment by deepakkumardk** [comment]
Any update on this issue as it's very crucial.
I am fetching items from realm asynchronously using livedata (as per given sample), there's no lag in fetching the data, but when I update the recyclerview using **notifiydatasetchanged** with the fetched data of size 400 items the **UI freezes** with black screen. Paging  would solve this.
Is there any workaround of this.

**Comment by Zhuinden** [comment]
@deepakkumardk if your UI freezes on `notifyDataSetChanged`, your RecyclerView probably is inside a `NestedScrollView` or has nested scrolling disabled, and therefore doesn't actually recycle.

**Comment by deepakkumardk** [comment]
@Zhuinden Yes, I am using the `NestedScrollView` to support the pagination and remove the autoscroll to top or in between them position in recyclerview on update same [issue#144](https://github.com/realm/realm-android-adapters/issues/144) as of **`RealmAndroidAdapters`** but I am not using this instead I am getting this issue in standard recyclerview.
I am just observing the db and loading the data from db only.
Is there any workaround to prevent both of this issues?

**Comment by Zhuinden** [comment]
You need to remove the NestedScrollView and whatever it's currently solving you have to most likely solve with view types.

**Comment by arunkumar9t2** [comment]
I just published [Compass](https://github.com/arunkumar9t2/compass) which provides set of APIs to integrate Realm with Jetpack Paging 3.

The API looks like this:

```kotlin
val pagedPersons = RealmQuery { where<Person>() }.asPagingItems()
```

Here `asPagingItems()` will internally maintain an active realm instance, lazily load items in response to paging lib callbacks in a dedicated worker thread and automatically cleans up once `Flow` collection is stopped.

It also provides a way to read only subset of data into memory with the overload:

```kotlin
val pagedPersonNames = RealmQuery { where<Person>() }.asPagingItems { it.name }
```

Example with Android ViewModel:

```kotlin
class MyViewModel: ViewModel() {

    val results = RealmQuery { where<Task>() }.asPagingItems().cachedIn(viewModelScope)
}
```

It can support transforms, caching and seperators from Paging lib as described [here](https://developer.android.com/topic/libraries/architecture/paging/v3-transform). Any feedback appreciated.
